docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,231 @@
+  * Do we need to handle more complicated structure for specifying endpoints? Perhaps a priority?
+  * Does this get versioned with the rest of the API or is it a separate thing?
+  * How do we say "trust system installed roots"?  It is possible users will configure the cluster with a publicly signed certificate.
+
+## Methods
+
+Now that we know *what* we want to get to the client, the question is how.  We want to do this in as secure a way possible (as there are cryptographic keys involved) without requiring a lot of overhead in terms of information that needs to be copied around.
+
+### Method: Out of Band
+
+The simplest way to do this would be to simply put this object in a file and copy it around.  This is more overhead for the user, but it is easy to implement and lets users rely on existing systems to distribute configuration.
+
+For the (currently being designed) `kubeadm` flow, the command line might look like:
+
+```
+kubeadm join --cluster-info-file=my-cluster.json
+```
+
+Note that TLS bootstrap (which establishes a way for a client to authenticate itself to the server) is a separate issue and has its own set of methods.  This command line may have a TLS bootstrap token (or config file) on the command line also.
+
+### Method: HTTPS endpoint
+
+If the ClusterInfo information is hosted in a trusted place of HTTPS you can just request it that way.  This will use the root certificates that are installed on the system.  It may or may not be appropriate based on the user's constraints.
+
+```
+kubeadm join --cluster-info-url="https://example/mycluster.json"
+```
+
+This is really a shorthand for someone doing something like (assuming we support stdin with `-`):
+
+```
+curl https://example.com/mycluster.json | kubeadm join --cluster-info-file=-
+```
+
+If the user requires some auth to the HTTPS server (to keep the ClusterInfo object private) that can be done in the curl command equivalent.  Or we could eventually add it to `kubeadm` directly.
+
+### Method: JWS Token
+
+There won’t always be a trusted external endpoint to talk to and transmitting the locator file out of band is a pain.  However, we want something a bit more secure than just hitting HTTP and trusting whatever we get back.  In this case, we assume we have the following:
+  * An address for at least one of the API servers (which will implement this API).
+    * This address is technically an HTTPS URL base but is often expressed as a bare domain or IP.
+  * A shared secret token
+
+An interesting aspect here is that this information is often easily obtained before the API server is configured or started.  This makes some cluster bring-up scenarios much easier.
+
+We put some structure into the token.  It has both a token identifier and the token value, separated by a dot.  Example:
+
+```
+A81E5d4DwI.0ok9tB1QhB
+```
+
+The first part of the token is the `token-id`.  The second part is the `token-secret`.
```

Comment:
I like the token IDs when it comes time to verify the JWS, this way we're not transmitting the secret we'll use to do that with the request itself, so that much less likely to get compromised data. 

kubeadm will hopefully just take the ID.TOKEN form directly copied from the master init command output, user shouldn't have to worry about splitting it up into ID or token, it's just a token to them.